File: ClientsToAnalyze/WinRT/2.1.5.1/Decomp/ProtoBuffer/Google.ProtocolBuffers/RpcUtil.cs
using System;

namespace Google.ProtocolBuffers;

public static class RpcUtil
{
	public static Action<T> SpecializeCallback<T>(Action<IMessage> action) where T : IMessage<T>
	{
		return delegate(T message)
		{
			action(message);
		};
	}

	public static Action<IMessage> GeneralizeCallback<TMessage, TBuilder>(Action<TMessage> action, TMessage defaultInstance) where TMessage : class, IMessage<TMessage, TBuilder> where TBuilder : IBuilder<TMessage, TBuilder>
	{
		return delegate(IMessage message)
		{
			TMessage val = message as TMessage;
			if (val == null)
			{
				val = defaultInstance.CreateBuilderForType().MergeFrom(message).Build();
			}
			action(val);
		};
	}
}

File: ClientsToAnalyze/Android/1.5.0.45-165.VoiceChat/app/src/main/java/com/google/protobuf/RpcUtil.java
package com.google.protobuf;

/* loaded from: classes.dex */
public final class RpcUtil {
    private RpcUtil() {
    }

    /* JADX WARN: Multi-variable type inference failed */
    public static <Type extends Message> RpcCallback<Type> specializeCallback(RpcCallback<Message> rpcCallback) {
        return rpcCallback;
    }

    public static <Type extends Message> RpcCallback<Message> generalizeCallback(final RpcCallback<Type> originalCallback, final Class<Type> originalClass, final Type defaultInstance) {
        return new RpcCallback<Message>() { // from class: com.google.protobuf.RpcUtil.1
            @Override // com.google.protobuf.RpcCallback
            public void run(Message parameter) {
                Message messageCopyAsType;
                try {
                    messageCopyAsType = (Message) originalClass.cast(parameter);
                } catch (ClassCastException e) {
                    messageCopyAsType = RpcUtil.copyAsType(defaultInstance, parameter);
                }
                originalCallback.run(messageCopyAsType);
            }
        };
    }

    /* JADX INFO: Access modifiers changed from: private */
    public static <Type extends Message> Type copyAsType(Type type, Message message) {
        return (Type) type.newBuilderForType().mergeFrom(message).build();
    }

    public static <ParameterType> RpcCallback<ParameterType> newOneTimeCallback(final RpcCallback<ParameterType> originalCallback) {
        return new RpcCallback<ParameterType>() { // from class: com.google.protobuf.RpcUtil.2
            private boolean alreadyCalled = false;

            @Override // com.google.protobuf.RpcCallback
            public void run(ParameterType parameter) {
                synchronized (this) {
                    if (this.alreadyCalled) {
                        throw new AlreadyCalledException();
                    }
                    this.alreadyCalled = true;
                }
                originalCallback.run(parameter);
            }
        };
    }

    public static final class AlreadyCalledException extends RuntimeException {
        private static final long serialVersionUID = 5469741279507848266L;

        public AlreadyCalledException() {
            super("This RpcCallback was already called and cannot be called multiple times.");
        }
    }
}
